File: backend/app/decision_engine.py
from __future__ import annotations

import math

from .config import DecisionPolicy
from .models import (
    DecisionCode,
    DecisionOutput,
    MotionState,
    Permission,
    RiskLevel,
    SafetyGate,
    TelemetryIn,
    TimingResult,
)


LABELS = {
    DecisionCode.STAY: "原地守望",
    DecisionCode.WATCH: "增强监测",
    DecisionCode.PREPARE: "建议准备迁移",
    DecisionCode.MIGRATE_NOW: "建议立即迁移",
    DecisionCode.VERIFY_ONLY: "只提醒，等待复核",
    DecisionCode.NO_GO: "No-Go：禁止迁移",
    DecisionCode.EMERGENCY_STOP: "异常停车并转人工",
}


def _timing(policy: DecisionPolicy, telemetry: TelemetryIn) -> tuple[TimingResult, float]:
    environment = telemetry.environment
    remaining_cm = policy.danger_water_level_cm - environment.water_level_cm
    if remaining_cm <= 0:
        time_to_threshold = 0.0
    elif environment.rise_rate_cm_min == 0:
        time_to_threshold = math.inf
    else:
        time_to_threshold = remaining_cm / environment.rise_rate_cm_min

    meters_per_minute = policy.max_speed_kmh * 1000 / 60
    route_time = policy.route_distance_m / meters_per_minute
    queue_batches = math.ceil(policy.queue_ahead / policy.batch_size)
    queue_time = queue_batches * policy.batch_interval_min
    latest_start = time_to_threshold - route_time - queue_time - policy.safety_buffer_min

    return (
        TimingResult(
            remaining_cm=remaining_cm,
            time_to_threshold_min=None if math.isinf(time_to_threshold) else time_to_threshold,
            route_time_min=route_time,
            queue_time_min=queue_time,
            latest_safe_start_min=None if math.isinf(latest_start) else latest_start,
        ),
        latest_start,
    )


def _gates(telemetry: TelemetryIn) -> list[SafetyGate]:
    vehicle = telemetry.vehicle
    site = telemetry.site
    return [
        SafetyGate(id="route-dry", label="路线全程干燥", passed=site.route_dry, detail="未发现水触点" if site.route_dry else "路线见水或触发禁行阈值"),
        SafetyGate(id="route-open", label="路线与出口可用", passed=not site.route_blocked, detail="路线与出口在线" if not site.route_blocked else "闸机、坡道或出口不可用"),
        SafetyGate(id="occupants", label="车内无人或宠物", passed=vehicle.occupants_clear, detail="乘员检测通过" if vehicle.occupants_clear else "检测到乘员或宠物"),
        SafetyGate(id="charging", label="充电枪已拔除", passed=vehicle.charging_disconnected, detail="充电互锁通过" if vehicle.charging_disconnected else "充电连接仍在"),
        SafetyGate(id="vehicle", label="车辆关键系统健康", passed=vehicle.vehicle_healthy, detail="关键系统自检通过" if vehicle.vehicle_healthy else "存在关键车辆故障"),
        SafetyGate(id="positioning", label="定位与场端在线", passed=vehicle.positioning_online, detail="定位与场端心跳正常" if vehicle.positioning_online else "定位或场端失联"),
        SafetyGate(id="network", label="主备通信可用", passed=vehicle.network_online, detail="主备链路在线" if vehicle.network_online else "通信链路不可用"),
        SafetyGate(id="operator", label="远程安全员在线", passed=vehicle.emergency_operator_online, detail="急停与人工接管可用" if vehicle.emergency_operator_online else "无法保证人工接管"),
        SafetyGate(id="water-contact", label="未触发水触禁行", passed=not vehicle.water_contact_triggered, detail="未触发水触" if not vehicle.water_contact_triggered else "车辆水触传感器已触发"),
    ]


def _risk(
    telemetry: TelemetryIn,
    policy: DecisionPolicy,
    latest_start: float,
    gates: list[SafetyGate],
) -> RiskLevel:
    if (
        telemetry.vehicle.water_contact_triggered
        or telemetry.environment.water_level_cm >= policy.danger_water_level_cm
        or latest_start <= 0
    ):
        return RiskLevel.CRITICAL
    if not all(gate.passed for gate in gates) or latest_start <= policy.migrate_horizon_min:
        return RiskLevel.HIGH
    if (
        latest_start <= policy.prepare_horizon_min
        or telemetry.environment.rainfall_mm_h >= policy.rain_watch_threshold_mm_h
    ):
        return RiskLevel.MEDIUM
    return RiskLevel.LOW


def evaluate_decision(
    telemetry: TelemetryIn,
    policy: DecisionPolicy | None = None,
    *,
    owner_authorized: bool = False,
) -> DecisionOutput:
    policy = policy or DecisionPolicy()
    timing, latest_start = _timing(policy, telemetry)
    gates = _gates(telemetry)
    failed_gates = [gate for gate in gates if not gate.passed]
    environment = telemetry.environment
    vehicle = telemetry.vehicle
    disagreement = abs(environment.water_level_cm - environment.secondary_water_level_cm)

    motion_fault = vehicle.motion_state == MotionState.MOVING and (
        not vehicle.positioning_online
        or not vehicle.network_online
        or not vehicle.emergency_operator_online
        or not vehicle.vehicle_healthy
        or vehicle.water_contact_triggered
    )

    if motion_fault:
        decision = DecisionCode.EMERGENCY_STOP
        permission = Permission.DENIED
        reason = "车辆移动期间安全链路异常，立即执行最小风险停车并转人工。"
    elif failed_gates:
        decision = DecisionCode.NO_GO
        permission = Permission.DENIED
        reason = "安全闸未全部通过：" + "、".join(gate.label for gate in failed_gates) + "。"
    elif environment.water_level_cm >= policy.danger_water_level_cm:
        decision = DecisionCode.NO_GO
        permission = Permission.DENIED
        reason = "当前水位已达到禁行阈值，禁止尝试涉水迁移。"
    elif (
        environment.sensor_confidence < policy.min_sensor_confidence
        or disagreement > policy.max_sensor_disagreement_cm
    ):
        decision = DecisionCode.VERIFY_ONLY
        permission = Permission.DENIED
        reason = (
            f"多源证据不足：置信度 {environment.sensor_confidence * 100:.0f}%，"
            f"水位交叉差 {disagreement:.1f} cm；系统只提醒，不下发移动权限。"
        )
    elif latest_start <= 0:
        decision = DecisionCode.NO_GO
        permission = Permission.DENIED
        reason = "最晚安全启动窗口已经关闭，禁止迟发迁移并转人工处置。"
    elif latest_start <= policy.migrate_horizon_min:
        decision = DecisionCode.MIGRATE_NOW
        permission = Permission.GRANTED if owner_authorized else Permission.AWAITING_OWNER
        reason = (
            f"最晚安全启动窗口仅剩 {latest_start:.1f} 分钟；"
            + ("车主单次授权已确认。" if owner_authorized else "仍需车主单次授权。")
        )
    elif latest_start <= policy.prepare_horizon_min:
        decision = DecisionCode.PREPARE
        permission = Permission.AWAITING_OWNER
        reason = f"最晚安全启动窗口剩余 {latest_start:.1f} 分钟，建议预登记高位点并准备授权。"
    elif (
        environment.rainfall_mm_h >= policy.rain_watch_threshold_mm_h
        or environment.rise_rate_cm_min >= 0.5
    ):
        decision = DecisionCode.WATCH
        permission = Permission.NONE
        reason = "强降雨或水位上涨较快，提升采样频率并持续计算安全窗口。"
    else:
        decision = DecisionCode.STAY
        permission = Permission.NONE
        reason = "风险仍低且安全窗口充足，保持原位并持续守望。"

    return DecisionOutput(
        decision=decision,
        label=LABELS[decision],
        risk_level=_risk(telemetry, policy, latest_start, gates),
        permission=permission,
        authorized_to_move=decision == DecisionCode.MIGRATE_NOW and owner_authorized,
        reason=reason,
        sensor_disagreement_cm=disagreement,
        timing=timing,
        safety_gates=gates,
    )

File: src/decision-engine.js
/**
 * XPENG HighGround 概念决策引擎
 *
 * 这是可解释、可测试的概念验证代码，不连接真实车辆，也不构成量产安全功能。
 * 所有车辆动作必须经过独立安全壳、车主单次授权和封闭场地验证。
 */

export const DECISIONS = Object.freeze({
  STAY: "STAY",
  WATCH: "WATCH",
  PREPARE: "PREPARE",
  MIGRATE_NOW: "MIGRATE_NOW",
  VERIFY_ONLY: "VERIFY_ONLY",
  NO_GO: "NO_GO",
  EMERGENCY_STOP: "EMERGENCY_STOP",
  SYSTEM_HOLD: "SYSTEM_HOLD",
});

export const DECISION_META = Object.freeze({
  [DECISIONS.STAY]: { label: "原地守望", tone: "safe", permission: "NONE" },
  [DECISIONS.WATCH]: { label: "增强监测", tone: "watch", permission: "NONE" },
  [DECISIONS.PREPARE]: { label: "建议准备迁移", tone: "warning", permission: "AWAITING_OWNER" },
  [DECISIONS.MIGRATE_NOW]: { label: "建议立即迁移", tone: "danger", permission: "AWAITING_OWNER" },
  [DECISIONS.VERIFY_ONLY]: { label: "只提醒，等待复核", tone: "warning", permission: "DENIED" },
  [DECISIONS.NO_GO]: { label: "No-Go：禁止迁移", tone: "danger", permission: "DENIED" },
  [DECISIONS.EMERGENCY_STOP]: { label: "异常停车并转人工", tone: "danger", permission: "DENIED" },
  [DECISIONS.SYSTEM_HOLD]: { label: "输入异常，系统保持", tone: "warning", permission: "DENIED" },
});

export const DEFAULT_INPUTS = Object.freeze({
  rainfallMmH: 35,
  waterLevelCm: 4,
  secondaryWaterLevelCm: 4,
  riseRateCmMin: 0.2,
  dangerWaterLevelCm: 22,
  rainWatchThresholdMmH: 50,
  minSensorConfidence: 0.72,
  sensorConfidence: 0.94,
  maxSensorDisagreementCm: 5,
  routeDistanceM: 260,
  maxSpeedKmh: 5,
  queueAhead: 2,
  batchSize: 3,
  batchIntervalMin: 0.7,
  safetyBufferMin: 3,
  prepareHorizonMin: 20,
  migrateHorizonMin: 7,
  routeDry: true,
  routeBlocked: false,
  occupantsClear: true,
  chargingDisconnected: true,
  vehicleHealthy: true,
  positioningOnline: true,
  networkOnline: true,
  emergencyOperatorOnline: true,
  ownerAuthorized: false,
  waterContactTriggered: false,
  motionState: "PARKED",
});

const NUMBER_KEYS = [
  "rainfallMmH",
  "waterLevelCm",
  "secondaryWaterLevelCm",
  "riseRateCmMin",
  "dangerWaterLevelCm",
  "rainWatchThresholdMmH",
  "minSensorConfidence",
  "sensorConfidence",
  "maxSensorDisagreementCm",
  "routeDistanceM",
  "maxSpeedKmh",
  "queueAhead",
  "batchSize",
  "batchIntervalMin",
  "safetyBufferMin",
  "prepareHorizonMin",
  "migrateHorizonMin",
];

function finiteNumber(value) {
  const number = Number(value);
  return Number.isFinite(number) ? number : Number.NaN;
}

export function normalizeInputs(raw = {}) {
  const normalized = { ...DEFAULT_INPUTS, ...raw };
  for (const key of NUMBER_KEYS) normalized[key] = finiteNumber(normalized[key]);
  normalized.queueAhead = Math.max(0, Math.floor(normalized.queueAhead));
  normalized.batchSize = Math.max(1, Math.floor(normalized.batchSize));
  return normalized;
}

export function calculateTiming(inputs) {
  const rise = Math.max(0, inputs.riseRateCmMin);
  const remainingCm = inputs.dangerWaterLevelCm - inputs.waterLevelCm;
  const timeToThresholdMin = remainingCm <= 0
    ? 0
    : rise === 0
      ? Number.POSITIVE_INFINITY
      : remainingCm / rise;

  const metersPerMinute = inputs.maxSpeedKmh * 1000 / 60;
  const routeTimeMin = metersPerMinute > 0
    ? inputs.routeDistanceM / metersPerMinute
    : Number.POSITIVE_INFINITY;
  const queueBatches = Math.ceil(inputs.queueAhead / inputs.batchSize);
  const queueTimeMin = queueBatches * inputs.batchIntervalMin;
  const latestSafeStartMin = timeToThresholdMin
    - routeTimeMin
    - queueTimeMin
    - inputs.safetyBufferMin;

  return {
    remainingCm,
    timeToThresholdMin,
    routeTimeMin,
    queueTimeMin,
    latestSafeStartMin,
  };
}

function check(id, label, ok, detail, severity = "critical") {
  return { id, label, ok: Boolean(ok), detail, severity };
}

export function evaluateSafetyGates(inputs) {
  const gates = [
    check("route-dry", "路线全程干燥", inputs.routeDry, inputs.routeDry ? "未发现水触点" : "路线见水或触发禁行阈值"),
    check("route-open", "路线与出口可用", !inputs.routeBlocked, inputs.routeBlocked ? "闸机、坡道或出口不可用" : "路线与出口在线"),
    check("occupants", "车内无人或宠物", inputs.occupantsClear, inputs.occupantsClear ? "乘员检测通过" : "检测到乘员或宠物"),
    check("charging", "充电枪已拔除", inputs.chargingDisconnected, inputs.chargingDisconnected ? "充电互锁通过" : "充电连接仍在"),
    check("vehicle", "车辆关键系统健康", inputs.vehicleHealthy, inputs.vehicleHealthy ? "关键系统自检通过" : "存在关键车辆故障"),
    check("positioning", "定位与场端在线", inputs.positioningOnline, inputs.positioningOnline ? "定位与场端心跳正常" : "定位或场端失联"),
    check("network", "主备通信可用", inputs.networkOnline, inputs.networkOnline ? "主备链路在线" : "通信链路不可用"),
    check("operator", "远程安全员在线", inputs.emergencyOperatorOnline, inputs.emergencyOperatorOnline ? "急停与人工接管可用" : "无法保证人工接管"),
    check("water-contact", "未触发水触禁行", !inputs.waterContactTriggered, inputs.waterContactTriggered ? "车辆水触传感器已触发" : "未触发水触", "critical"),
  ];
  return {
    gates,
    allPassed: gates.every((gate) => gate.ok),
    failed: gates.filter((gate) => !gate.ok),
  };
}

function invalidInputReasons(inputs) {
  const reasons = [];
  for (const key of NUMBER_KEYS) {
    if (!Number.isFinite(inputs[key])) reasons.push(`${key} 不是有效数字`);
  }
  if (inputs.rainfallMmH < 0) reasons.push("降雨强度不能为负数");
  if (inputs.waterLevelCm < 0 || inputs.secondaryWaterLevelCm < 0) reasons.push("水位不能为负数");
  if (inputs.dangerWaterLevelCm <= 0) reasons.push("禁行水位必须大于 0");
  if (inputs.maxSpeedKmh <= 0 || inputs.maxSpeedKmh > 5) reasons.push("概念场景车速必须在 0–5 km/h 内");
  if (inputs.sensorConfidence < 0 || inputs.sensorConfidence > 1) reasons.push("传感器置信度必须在 0–1 内");
  return reasons;
}

function riskLevel(inputs, timing, safety) {
  if (inputs.waterContactTriggered || inputs.waterLevelCm >= inputs.dangerWaterLevelCm || timing.latestSafeStartMin <= 0) return "CRITICAL";
  if (!safety.allPassed || timing.latestSafeStartMin <= inputs.migrateHorizonMin) return "HIGH";
  if (timing.latestSafeStartMin <= inputs.prepareHorizonMin || inputs.rainfallMmH >= inputs.rainWatchThresholdMmH) return "MEDIUM";
  return "LOW";
}

function makeResult(inputs, timing, safety, decision, reason, extra = {}) {
  const meta = DECISION_META[decision];
  const authorizedToMove = decision === DECISIONS.MIGRATE_NOW
    && safety.allPassed
    && inputs.ownerAuthorized;
  return {
    decision,
    label: meta.label,
    tone: meta.tone,
    permission: authorizedToMove ? "GRANTED" : meta.permission,
    authorizedToMove,
    reason,
    riskLevel: riskLevel(inputs, timing, safety),
    inputs,
    timing,
    safety,
    ...extra,
  };
}

/**
 * 决策顺序体现“安全壳优先”：输入有效性 → 移动中异常 → 证据一致性
 * → 物理禁行条件 → 最晚安全启动时间 → 预警守望。
 */
export function evaluateDecision(rawInputs = {}) {
  const inputs = normalizeInputs(rawInputs);
  const invalidReasons = invalidInputReasons(inputs);
  const safety = evaluateSafetyGates(inputs);
  const timing = calculateTiming(inputs);

  if (invalidReasons.length) {
    return makeResult(inputs, timing, safety, DECISIONS.SYSTEM_HOLD, invalidReasons.join("；"), { invalidReasons });
  }

  const motionFault = inputs.motionState === "MOVING" && (
    !inputs.positioningOnline
    || !inputs.networkOnline
    || !inputs.emergencyOperatorOnline
    || !inputs.vehicleHealthy
    || inputs.waterContactTriggered
  );
  if (motionFault) {
    return makeResult(inputs, timing, safety, DECISIONS.EMERGENCY_STOP, "车辆移动期间安全链路异常，立即执行最小风险停车并转人工。");
  }

  if (!safety.allPassed) {
    return makeResult(
      inputs,
      timing,
      safety,
      DECISIONS.NO_GO,
      `安全闸未全部通过：${safety.failed.map((gate) => gate.label).join("、")}。`,
      { sensorDisagreementCm: Math.abs(inputs.waterLevelCm - inputs.secondaryWaterLevelCm) },
    );
  }

  if (inputs.waterLevelCm >= inputs.dangerWaterLevelCm) {
    return makeResult(inputs, timing, safety, DECISIONS.NO_GO, "当前水位已达到禁行阈值，禁止尝试涉水迁移。", { sensorDisagreementCm: Math.abs(inputs.waterLevelCm - inputs.secondaryWaterLevelCm) });
  }

  const sensorDisagreementCm = Math.abs(inputs.waterLevelCm - inputs.secondaryWaterLevelCm);
  const evidenceUncertain = inputs.sensorConfidence < inputs.minSensorConfidence
    || sensorDisagreementCm > inputs.maxSensorDisagreementCm;
  if (evidenceUncertain) {
    return makeResult(
      inputs,
      timing,
      safety,
      DECISIONS.VERIFY_ONLY,
      `多源证据不足：置信度 ${(inputs.sensorConfidence * 100).toFixed(0)}%，水位交叉差 ${sensorDisagreementCm.toFixed(1)} cm；系统只提醒，不下发移动权限。`,
      { sensorDisagreementCm },
    );
  }

  if (timing.latestSafeStartMin <= 0) {
    return makeResult(inputs, timing, safety, DECISIONS.NO_GO, "最晚安全启动窗口已经关闭，禁止迟发迁移并转人工处置。", { sensorDisagreementCm });
  }

  if (timing.latestSafeStartMin <= inputs.migrateHorizonMin) {
    const ownerText = inputs.ownerAuthorized ? "车主单次授权已确认。" : "仍需车主单次授权。";
    return makeResult(
      inputs,
      timing,
      safety,
      DECISIONS.MIGRATE_NOW,
      `最晚安全启动窗口仅剩 ${Math.max(0, timing.latestSafeStartMin).toFixed(1)} 分钟；${ownerText}`,
      { sensorDisagreementCm },
    );
  }

  if (timing.latestSafeStartMin <= inputs.prepareHorizonMin) {
    return makeResult(inputs, timing, safety, DECISIONS.PREPARE, `最晚安全启动窗口剩余 ${timing.latestSafeStartMin.toFixed(1)} 分钟，建议预登记高位点并准备授权。`, { sensorDisagreementCm });
  }

  if (inputs.rainfallMmH >= inputs.rainWatchThresholdMmH || inputs.riseRateCmMin >= 0.5) {
    return makeResult(inputs, timing, safety, DECISIONS.WATCH, "强降雨或水位上涨较快，提升采样频率并持续计算安全窗口。", { sensorDisagreementCm });
  }

  return makeResult(inputs, timing, safety, DECISIONS.STAY, "风险仍低且安全窗口充足，保持原位并持续守望。", { sensorDisagreementCm });
}

export function formatMinutes(value) {
  if (value === Number.POSITIVE_INFINITY) return "充足";
  if (!Number.isFinite(value)) return "--";
  return `${Math.max(0, value).toFixed(1)} min`;
}
